### PR TITLE
Add ability to set attributions in WIT

### PR DESCRIPTION
**Pull request by jameswex**
* Motivation for features / changes

Enabling WIT to be able to display feature-level attributions for individual predictions if the model provides it.

* Technical description of changes

- Update visuals of attribution(saliency) viewing in vz-example-viewer, including better layout/coloring, and removing some unnecessary controls.
- Update inference_utils.py logic to be able to handle inference results with new, optional attribution values.
- Update WitWidget custom_predict_functions to be able to optionally return attributions in addition to predictions, by returning a dict containing both. Update parsing of those results to handle that case.
- Update AI platform custom predict function to pass any attributions it generates back through that newly-defined dict.
- Update WitWidget colab and jupyter code to pass those optional attributions to javascript when inferences are made.
- Update WIT element to handle attributions passed to it by updating the visdata for the elements that have attribution information to include that information so it is visualized by vz-example-viewer, and usable in Facets Dive.

* Screenshots of UI changes

attribs

* Detailed steps to verify changes work correctly (as executed by you)

Created custom predict function that created fake attributions and inferences, tested it in colab and jupyter notebooks to see correct visual handling of those attributions.

* Alternate designs / implementations considered

### CODE REVIEW

**Comment by jameswex** [comment]
@tolga-b please review

File: tensorboard/plugins/interactive_inference/witwidget/notebook/base.py [hunk 1]
```diff
@@ -122,8 +123,11 @@ def infer_impl(self):
       self.estimator_and_spec.get('estimator'),
       self.estimator_and_spec.get('feature_spec'),
       self.custom_predict_fn)
-    infer_objs.append(inference_utils.run_inference_for_inference_results(
+    (preidctions, attributions) = (
```
Comment (tolga-b) [inline tensorboard/plugins/interactive_inference/witwidget/notebook/base.py]: typo

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py [hunk 2]
```diff
@@ -413,6 +413,24 @@ def set_custom_predict_fn(self, predict_fn):
       - For regression: A 1D list of numbers, with a regression score for each
         example being predicted.
 
+    Optionally, if attributions can be returned by the model with each
+    prediction, then this method can return a dict with the key 'predictions'
+    containing the preidctions result list described above, and with the key
```
Comment (tolga-b) [inline tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: typo preidctions, this happens in a lot of places so mass replace should fix it.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 3]
```diff
@@ -4266,6 +4287,42 @@ <h2>Show similarity to selected datapoint</h2>
...
       },
 
+      newAttributions_: function(attributions) {
+        if (Object.keys(attributions).length == 0) {
+          return;
+        }
+        for (let i = 0; i < attributions.indices.length; i++) {
+          const idx = attributions.indices[i];
+          const datapoint = this.visdata[idx];
+          // For now, we only display attribution from the first model, if WIT
```
Comment (tolga-b) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: Add TODO here for two models

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 4]
```diff
@@ -2186,6 +2186,27 @@ <h2>Show similarity to selected datapoint</h2>
           observer: 'newInferences_',
           value: () => ({})
         },
+        // Attributions from inference. A dict with two fields: 'indices' and
+        // 'attributions'. Inidicies contains a list of example indices that
```
Comment (tolga-b) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: typo Inidicies

File: tensorboard/plugins/interactive_inference/witwidget/notebook/base.py (hunk 1 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/witwidget/notebook/base.py]: done

File: tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py (hunk 2 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/witwidget/notebook/visualization.py]: done

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 3 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 4 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done

File: tensorboard/components/vz_example_viewer/vz-example-viewer.ts [hunk 5]
```diff
@@ -397,7 +397,7 @@ Polymer({
           (d: {}, i: number) => this.getColorForSaliency(val[i]) :
           () => this.getColorForSaliency(val);
       this.selectAll(
-            `input.${this.sanitizeFeature(feat.name)}.value-pill`)
+            `iron-autogrow-textarea.${this.sanitizeFeature(feat.name)}.value-pill`)
```
[the remark this answers is not in the document]
Comment (jameswex) [reply tensorboard/components/vz_example_viewer/vz-example-viewer.ts]: This code actually works correctly without the selector for the element this, so i removed it. Thanks for catching this.

**Comment by jameswex** [comment]
@stephanwlee please review thx

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 6]
```diff
@@ -2186,6 +2186,27 @@ <h2>Show similarity to selected datapoint</h2>
...
+        //   values of a multivalent feature - such as pixels in an image or
+        //   numbers in a list of numbers.
+        // - A 2D list for sparse attributions of a multivalent feature. The
+        //   first list contains the feature values for which attributions
+        //   exist. The second list contains the numeric attributions for each
+        //   of those feature values.
+        attributions: {
+          type: Object,
+          observer: 'newAttributions_',
+          value: () => ({})
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: `null` might be better.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 7]
```diff
@@ -4266,6 +4287,43 @@ <h2>Show similarity to selected datapoint</h2>
...
+                Array.isArray(attribs[keys[j]][0])) {
+              const attribSparseKeyKey = attributionPrefix + keys[j] +
+                attributionSparseKeySuffix;
+              datapoint[attribSparseKeyKey] = attribs[keys[j]][0];
+              const attribSparseValueKey = attributionPrefix + keys[j] +
+                attributionSparseValueSuffix;
+              datapoint[attribSparseValueKey] = attribs[keys[j]][1];
+            } else {
+              const attribKey = attributionPrefix + keys[j];
+              datapoint[attribKey] = attribs[keys[j]];
```
Comment (stephanwlee) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: it feels un-Polymermic to mutate data like this and then trigger a draw (as opposed to using prop changes to trigger data binding). Can you consider creating another object and do ```this.set(`visdata[${index}]`, newObject)```?

File: tensorboard/plugins/interactive_inference/interactive_inference_plugin.py [hunk 8]
```diff
@@ -282,8 +282,9 @@ def _infer(self, request):
             request.args.get('use_predict') == 'true',
             request.args.get('predict_input_tensor'),
             request.args.get('predict_output_tensor'))
-        infer_objs.append(inference_utils.run_inference_for_inference_results(
-          examples_to_infer, serving_bundle))
+        (predictions, _) = inference_utils.run_inference_for_inference_results(
+          examples_to_infer, serving_bundle)
```
[the remark this answers is not in the document]
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/interactive_inference_plugin.py]: done

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 7 above)
Comment (jameswex) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: fixing

**Pull request merged**